### PR TITLE
Make StorageImpl member variables largely private and use getters and setters

### PR DESCRIPTION
**Pull request by cpuhrsch**

### CODE REVIEW

File: aten/src/ATen/StorageImpl.h [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include <ATen/Retainable.h>
 #include <TH/THTypeConversion.hpp>
 #include <atomic>
+#include <stdexcept>
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.h]: What's this?

Comment (cpuhrsch) [reply aten/src/ATen/StorageImpl.h]: Hm, I must have had a throw in there at some point. I'll remove this.

File: aten/src/ATen/StorageImpl.h [hunk 2]
```diff
@@ -41,29 +42,25 @@ namespace at {
...
-  at::DataPtr data_ptr;
-  ptrdiff_t size;
-  bool resizable;
-  at::Allocator* allocator;
-  std::unique_ptr<THFinalizer> finalizer;
   StorageImpl(StorageImpl&) = delete;
   StorageImpl(const StorageImpl&) = delete;
-  StorageImpl(StorageImpl&&) = delete;
+  // NB: Don't move ref count!
+  StorageImpl(StorageImpl&& other) = delete;
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.h]: FWIW, the default move constructor won't work anyway, because std::atomic is not movable.

**Comment by cpuhrsch** [comment]
I'm going to try applying the suggested changes, but more likely than not i'll need to close this and do one member variable at a time.

File: aten/src/ATen/StorageImpl.h [hunk 3]
```diff
@@ -74,42 +70,72 @@ struct AT_API StorageImpl : public Retainable {
...
   }
+
+ private:
+  at::ScalarType scalar_type_;
+  at::DataPtr data_ptr_;
+  ptrdiff_t size_;
+  bool resizable_;
+
+ public:
+  at::Allocator* allocator_;
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.h]: You'll make these private soon?

File: aten/src/ATen/THLongStorageView.h [hunk 4]
```diff
@@ -53,22 +53,30 @@ class THLongStorageView {
...
 
-    if(zero_dim_to_one && ref.size() == 0) {
+    if (zero_dim_to_one && ref.size() == 0) {
       // make storage of size 0 actually a 1-length storage with 1 element
       // so that our 0-dim tensors get allocated as 1-dim inside TH
+
       one = 1;
-      storage.data_ptr = {&one, kCPU}; // non-owning
-      storage.size = 1;
+      storage = new StorageImpl(
```
Comment (ezyang) [inline aten/src/ATen/THLongStorageView.h]: I found your leak, it's here. storage is a raw pointer, and you aren't explicitly freeing it, so you're leaking these StorageImpl structs. You should store a `Storage` instead to avoid the leak.

File: aten/src/ATen/THLongStorageView.h [hunk 5]
```diff
@@ -53,22 +53,30 @@ class THLongStorageView {
...
+          at::CTypeToScalarType<th::from_type<int64_t>>::to(),
```
Comment (ezyang) [inline aten/src/ATen/THLongStorageView.h]: I think the `th::from_type` here is not needed.

File: aten/src/ATen/THLongStorageView.h (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/THLongStorageView.h]: (Storing a `Storage` pointery thing here is still suboptimal, because the lifetime of pointers like `&one` is tied to the lifetime of THLongStorageView, so if storage lives on beyond this class you'll still corrupt memory. But I guess since the idea is to get rid of this class entirely I can live with it... as long as there's a comment.)

(I don't know why you stopped storing StorageImpl in this class directly, but I guess you had your reasons?)

File: aten/src/ATen/THLongStorageView.h [hunk 6]
```diff
@@ -53,22 +53,30 @@ class THLongStorageView {
...
+          1,
+          {&one, kCPU}, // non-owning
+          getTHDefaultAllocator(),
+          false);
     } else {
-      storage.data_ptr = {const_cast<void*>(static_cast<const void*>(ref.data())), kCPU}; // non-owning
-      storage.size = ref.size();
+      storage = new StorageImpl(
```
Comment (ezyang) [inline aten/src/ATen/THLongStorageView.h]: leak here too

File: aten/src/ATen/THLongStorageView.h [hunk 7]
```diff
@@ -53,22 +53,30 @@ class THLongStorageView {
...
       // so that our 0-dim tensors get allocated as 1-dim inside TH
+
       one = 1;
-      storage.data_ptr = {&one, kCPU}; // non-owning
-      storage.size = 1;
+      storage = new StorageImpl(
+          at::CTypeToScalarType<th::from_type<int64_t>>::to(),
+          1,
+          {&one, kCPU}, // non-owning
+          getTHDefaultAllocator(),
```
Comment (ezyang) [inline aten/src/ATen/THLongStorageView.h]: I don't think these should get an allocator, because they are not "reallocatable" in the ordinary sense.

File: aten/src/TH/generic/THStorage.cpp [hunk 8]
```diff
@@ -137,25 +137,34 @@ void THStorage_(resize)(THStorage *storage, ptrdiff_t size)
...
+  std::swap(storage1->scalar_type(), storage2->scalar_type());
+  std::swap(storage1->data_ptr(), storage2->data_ptr());
+  ptrdiff_t tmp_size = storage1->size();
+  storage1->set_size(storage2->size());
+  storage2->set_size(tmp_size);
+  bool tmp_bool = storage1->resizable();
+  storage1->set_resizable(storage2->resizable());
+  storage2->set_resizable(tmp_bool);
+  std::swap(storage1->allocator_, storage2->allocator_);
+  std::swap(storage1->finalizer_, storage2->finalizer_);
```
Comment (ezyang) [inline aten/src/TH/generic/THStorage.cpp]: I know I gave conflicting advice here, but I think my preferred implementations of swap is using a hypothetical move constructor on storage. The move constructor is closer to the fields and more likely to be updated if someone adds another field. But this is OK for now.

**Review by ezyang — CHANGES_REQUESTED** [review]
leaks memory

File: aten/src/TH/generic/THStorage.cpp (hunk 8 above)
Comment (cpuhrsch) [reply aten/src/TH/generic/THStorage.cpp]: Makes sense. I was planning on doing something like that in an upcoming PR as this is becoming too large.

File: aten/src/ATen/StorageImpl.h (hunk 3 above)
Comment (cpuhrsch) [reply aten/src/ATen/StorageImpl.h]: Yes, in an upcoming PR to keep this one small.

File: aten/src/ATen/THLongStorageView.h (hunk 4 above)
Comment (cpuhrsch) [reply aten/src/ATen/THLongStorageView.h]: Thanks! Great find :) - I used a pointer, because the assignment constructor for StorageImpl is deleted and I didn't want to implement it.

File: aten/src/ATen/THLongStorageView.h (hunk 5 above)
Comment (cpuhrsch) [reply aten/src/ATen/THLongStorageView.h]: I'll keep it since we'll delete this whole thing anyway.

File: aten/src/ATen/Storage.h [hunk 9]
```diff
@@ -21,6 +21,9 @@ struct AT_API Storage {
   Storage(const Storage&) = delete;
   Storage(Storage&&) = delete;
   Storage(const Storage&&) = delete;
+  void set_pImpl(StorageImpl* storage_impl) {
+    storage_impl_ = storage_impl;
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: This looks like a dangerous function. Is it really needed?

Comment (ezyang) [reply aten/src/ATen/Storage.h]: OK, looks like it's just used by THLongStorageView, which is going away. If you want to be nice add a comment saying when we can delete this function, but since we plan to get rid of this soon it's not totally necessary.

File: aten/src/ATen/THLongStorageView.h [hunk 10]
```diff
@@ -53,22 +53,33 @@ class THLongStorageView {
...
 
-    if(zero_dim_to_one && ref.size() == 0) {
+    if (zero_dim_to_one && ref.size() == 0) {
       // make storage of size 0 actually a 1-length storage with 1 element
       // so that our 0-dim tensors get allocated as 1-dim inside TH
+
       one = 1;
-      storage.data_ptr = {&one, kCPU}; // non-owning
-      storage.size = 1;
+      storage.set_pImpl(new StorageImpl(
```
Comment (ezyang) [inline aten/src/ATen/THLongStorageView.h]: This comment is not requesting a code change, it's just a "for your information."

Because this is a constructor, you could have avoided the need for `set_pImpl` by simply calling the correct constructor on `storage` at the very beginning. Obviously the `StorageImpl` to allocate in this case is somewhat complex, so you'd add a helper function to actually compute and return it.

File: aten/src/TH/generic/THStorage.cpp (hunk 8 above)
Comment (ezyang) [reply aten/src/TH/generic/THStorage.cpp]: This function is now screaming to be implemented as a method (or, at very least, a friend function.)

```
  bool tmp_bool = storage1->resizable();
  storage1->set_resizable(storage2->resizable());
  storage2->set_resizable(tmp_bool);
```

is terrible and we shouldn't encourage it.

**Review by ezyang — APPROVED** [review]
There are some hokey bits but I'm OK leaving them for follow ups.

File: aten/src/TH/generic/THTensor.cpp [hunk 11]
```diff
@@ -596,6 +596,20 @@ int THTensor_(isContiguous)(const THTensor *self)
   return 1;
 }
 
+int THTensor_(isSize)(const THTensor *self, const THLongStorage *dims)
```
Comment (gchanan) [inline aten/src/TH/generic/THTensor.cpp]: why is this here?  It was removed in https://github.com/pytorch/pytorch/pull/10146 -- your merge probably went awry.

File: aten/src/THC/generic/THCTensor.cpp [hunk 12]
```diff
@@ -523,6 +523,20 @@ int THCTensor_(isContiguous)(THCState *state, const THCTensor *self)
   return THCTensor_isContiguous(state, self);
 }
 
+int THCTensor_(isSize)(THCState *state, const THCTensor *self, const THLongStorage *dims)
```
Comment (gchanan) [inline aten/src/THC/generic/THCTensor.cpp]: here too.
